added session to udp

y0sher opened the pull request:
still playing with automation testing.. hence the draft

y0sher committed:
added session to udp

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// GetIfExist gets a session from the cache based on address
+func (s *sessionCache) GetIfExist(addr string) (NetworkSession, error) {
```
Why `addr` is if type `string` and not `net.Addr`?

zalmen replied:
This comment is also relevant for the choice to use `string` as the type of the key in the `sessionCache` struct

y0sher replied:
`net.Addr` is an interface, `UDPAddr` is passed by pointer hence comparison wouldn't work using the bare structs. `UDPAddr` is composed of byte array for IP and integer for the port. didn't seem like a big difference for me and was less prone to pointer-interface errors.

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+
+// GetOrCreate get's a session if it exists, or try to init a new one based on the address and key given.
+// *NOTE*: use of dialFunc might actually trigger message sending.
+func (s *sessionCache) GetOrCreate(remote string, pubKey p2pcrypto.PublicKey) (NetworkSession, error) {
```
`remote` should be of type `net.Addr`

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+		s.expire()
+	}
+
+	session, err := s.dialFunc(remote, pubKey)
```
`dialFunc` might take a long time, you need to avoid from holding the mutex locked while this method is called

y0sher replied:
this is not entirely true, dialFunc would create a session (should be a cheap, short cryptographic action) and then send a UDP message which again should be cheap and short.

zalmen replied:
Sending a message is not a short op, compared to memory/cpu ops

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+// simple cache for storing sessions
+const maxSessions = 1024
+
+type storedSession struct {
```
How is this struct different than the ConnectionPool?

y0sher replied:
In fact they are quite similar, connectionPool is more complicated because it has to synchronize long dials. also connection pool sessions are cached by the PublicKey received in session key exchange. for UDP we can't do this unless we attach the pubkey to each message.

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+	return session, nil
+}
+
+func (s *sessionCache) handleIncoming(from string, ns NetworkSession) {
```
replace `string` with `net.Addr`

zalmen commented on p2p/net/udp.go:
```diff
@@ -81,14 +90,76 @@ func newUDPListener(listenAddress *net.UDPAddr) (*net.UDPConn, error) {
+	}
+
+	session := createSession(n.local.PrivateKey(), remote)
+	handshakeMessage, err := generateHandshakeMessage(session, n.config.NetworkID, 1010, n.local.PublicKey())
```
1010? Isn't it defined in the configuration?

zalmen replied:
Also, is it the same handshake message that will be sent over a tcp session?

y0sher replied:
you're right about the configuration and port. I'll fix that. in fact, the port is not used in the message right now since we exchange it in the "Ping" message.
same handshake message for now indeed. when we'll want to optimize discovery we can tailor the handshake message to also send discovery related stuff. but the key-exchange is the same.

zalmen commented on p2p/net/udp.go:
```diff
@@ -81,14 +90,76 @@ func newUDPListener(listenAddress *net.UDPAddr) (*net.UDPConn, error) {
+	if err != nil {
+		return nil, err
+	}
+	_, err = n.conn.WriteToUDP(handshakeMessage, raddr)
```
Do we need to wait for the handshake response? Can the local node start send data after it sent the handshake request, but before it received the response?

zalmen replied:
If the answer is no, where do you store the fact that the session is waiting for an HS response

y0sher replied:
Currently, the session we're using doesn't wait for a response. the initiator knows both the keys and creates a shared secret with it. when the receiver receives the message with the initiator public key, he can create a shared secret with it and his own key.

zalmen commented on p2p/net/udp.go:
```diff
@@ -81,14 +90,76 @@ func newUDPListener(listenAddress *net.UDPAddr) (*net.UDPConn, error) {
+	return session, nil
+}
+
+func (n *UDPNet) handleSession(message []byte) (NetworkSession, error) {
```
`handleSession` doesn't say much on what the purpose of this method and when it should be called. Rename it or add documentation

y0sher committed:
public keys are now carried in each UDP message header. this is also true for NetworkID and ClientVersion.
as we can't track when UDP clients update version or change keys. we need every message to include this basic info. the public key is used to identify or create a new shared secret for secured&encrypted communications.

y0sher committed:
fmt

y0sher committed:
remove unused code

zalmen commented on p2p/gossip/protocol.go:
```diff
@@ -261,15 +262,16 @@ func (prot *Protocol) removePeer(peer p2pcrypto.PublicKey) {
-	if payload := msg.GetPayload(); payload != nil {
-		data = service.DataBytes{Payload: payload}
-	} else if wrap := msg.GetMsg(); wrap != nil {
+	fmt.Println("############ WTF ", msg.Payload)
```
remove

zalmen commented on p2p/gossip/protocol.go:
```diff
@@ -261,15 +262,16 @@ func (prot *Protocol) removePeer(peer p2pcrypto.PublicKey) {
 	}
 
 	protocol := msg.Metadata.NextProtocol
+	fmt.Printf("############# CALCING %v proto with %v ############\r\n", protocol, data.Bytes())
```
remove

zalmen commented on p2p/net/session_cache.go:
```diff
@@ -0,0 +1,90 @@
+)
+
+// simple cache for storing sessions
+const maxSessions = 1024
```
This value has a major implications on the behavior of the node, it should be in the configuration

zalmen commented on p2p/net/udp.go:
```diff
@@ -81,14 +89,39 @@ func newUDPListener(listenAddress *net.UDPAddr) (*net.UDPConn, error) {
 	}
 
+	sealed := ns.SealMessage(data)
+	final := p2pcrypto.PrependPubkey(sealed, n.local.PublicKey()) // todo: prepend verison+networkid
```
remove todo

zalmen commented on p2p/udp.go:
```diff
@@ -135,15 +137,15 @@ func (mux *UDPMux) sendMessageImpl(peerPubkey p2pcrypto.PublicKey, protocol stri
 	switch x := payload.(type) {
 	case service.DataBytes:
-		protomessage.Data = &pb.ProtocolMessage_Payload{Payload: x.Bytes()}
+		protomessage.Payload = &pb.Payload{Data: &pb.Payload_Payload{x.Bytes()}}
```
use `createPayload`

y0sher committed:
cr comments
